### DISCUSSION TITLE
Issues with Fw::Buffers in Unit Tests

### DISCUSSION THREAD

**Discussion by ebenezerarunkumar** [nasa/fprime#1911] [Q&A] [answered]
Hello,

I am working on writing a unit test for a component(fprime v3.1.1) and I'm seeing some weird issues when serializing data into a Fw::Buffer.

A snipped of my code is shown below:

```
    Fw::Buffer data(nullptr, 32);
    Fw::SerializeStatus status = data.getSerializeRepr().serialize(apid);
    data.setSize(sizeof(apid));
    FW_ASSERT(status == Fw::FW_SERIALIZE_OK, status);
```

The code asserts with status 2 which corresponds to  `FW_SERIALIZE_NO_ROOM_LEFT,  //!< No room left in the buffer to serialize data` even when `apid` is a `U16`(2 bytes)...

This error happens regardless of whether I am serializing data into a buffer in my `Tester.cpp` file or my component's cpp file.

Does anyone have any idea why this might be happening?

**Comment by ebenezerarunkumar** [comment, marked as answer]
Have I made a mistake in how I am serializing data into a buffer?

**Reply by timcanham** [reply]
Without running the test code, I see one issue:

```
Fw::Buffer data(nullptr, 32);
```

`Fw::Buffer` requires real memory to back it, not a null pointer. Without a non-null pointer, the underlying serialization buffer initializes to size 0.

**Reply by ebenezerarunkumar** [reply]
Yes this makes sense ofc! I changed the first line to this and it now works!

``` 
U8 *const dataArr = new U8[100];
memset(dataArr, 0, 100);
Fw::Buffer data(dataArr, 100);
```

**Reply by timcanham** [reply]
Good! Now just make sure you delete the memory when the `Fw::Buffer` is out of scope. :grinning: A typical F Prime pattern is to use `Svc/BufferManager` to allocate and manage memory buffers. You ask it for an `Fw::Buffer` with the right amount of memory, and then return it when you are done. It avoids dynamic memory allocation while cycling software. 

https://nasa.github.io/fprime/UsersGuide/best/dynamic-memory.html

**Reply by ebenezerarunkumar** [reply]
Yup, sounds good! We use this pattern for our actual component cpp, but I just needed to create a buffer in my unit test Tester.cpp file! I appreciate the quick responses!
